lib/emerald/parser.rb

```diff
@@ -6,17 +6,27 @@ def initialize(source)
 
     def parse
       ast = []
-      result = parse_node(@source)
+      parse_lists_and_nodes(@source, ast)
+    end
+
+    private
+
+    def parse_lists_and_nodes(source, ast)
+      result = parse_node(source)
       while result
         node, source = result
-        ast.push(node) if node
+        if node.is_a? List
```

Comment:
The list is a node, so we shouldn't need to modify the list object after it is returned from the parse_list function, it should return fully parsed with its children. This top level method only needs to know what to parse, not how to.